ouroboros-consensus/test/consensus-test/Test/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,313 @@
+          ]
+          $ property True
+  let failure desc =
+        counterexample desc $
+          property False
+  -- Now check that the real implementation agrees with the model
+  case isPerasVotingAllowed pvv of
+    Right voteReason
+      | shouldVote ->
+          ok $ "VoteReason(" <> explainShallow voteReason <> ")"
+      | otherwise ->
+          failure $ "Expected not to vote, but got: " <> show voteReason
+    Left noVoteReason
+      | not shouldVote ->
+          ok $ "NoVoteReason(" <> explainShallow noVoteReason <> ")"
+      | otherwise ->
+          failure $ "Expected to vote, but got: " <> show noVoteReason
+
+{-------------------------------------------------------------------------------
+  Arbitrary helpers
+-------------------------------------------------------------------------------}
+
+-- * Peras parameters
+
+-- NOTE: we use a geometric distribution to bias towards smaller values.
+-- This increases the chance of covering all the voting rules more evenly,
+-- while still allowing for larger values to be generated occasionally.
+--
+-- Moreover, geometric(0.5) + 1 means that:
+--  - 50% chance of being 1
+--  - 25% chance of being 2
+--  - 12.5% chance of being 3
+--  ... and so on
+genPerasParams :: Gen PerasParams
+genPerasParams = do
+  _L <- fromIntegral . (+ 1) <$> geometric 0.5
+  _X <- fromIntegral . (+ 1) <$> geometric 0.5
+  _R <- fromIntegral . (+ 1) <$> geometric 0.5
+  _K <- fromIntegral . (+ 1) <$> geometric 0.5
+  pure
+    mkPerasParams
+      { perasBlockMinSlots = PerasBlockMinSlots _L
+      , perasCertArrivalThreshold = PerasCertArrivalThreshold _X
+      , perasIgnoranceRounds = PerasIgnoranceRounds _R
+      , perasCooldownRounds = PerasCooldownRounds _K
+      }
+
+-- * Slot numbers
+
+genSlotNo :: Gen SlotNo
+genSlotNo = do
+  n <- arbitrary
+  pure (SlotNo n)
+
+-- * Peras round numbers
+
+genPerasRoundNo :: Gen PerasRoundNo
+genPerasRoundNo = do
+  n <- arbitrary
+  pure (PerasRoundNo n)
+
+-- * Mocked certificate type
+
+-- NOTE: we could also use the real 'WithArrivalTime (ValidatedPerasCert blk)'
+-- here. However, this one is much easier to derive a 'Function' instance for,
+-- so we can actually generate the methods needed by 'PerasVotingView'.
+
+data TestCert
+  = TestCert
+  { tcArrivalTime :: RelativeTime
+  , tcRoundNo :: PerasRoundNo
+  }
+  deriving (Show, Eq, Generic)
+
+instance HasPerasCertRound TestCert where
+  getPerasCertRound = tcRoundNo
+
+-- | Generate a test certificate
+--
+-- NOTE: to improve the probabilities of covering all the paths in the code,
+-- we generate certificates relative to a given Peras round (the current one).
+genTestCert :: PerasRoundNo -> Gen TestCert
+genTestCert roundNo = do
+  arrivalTime <- RelativeTime <$> genNominalDiffTime50Years
+  offset <- choose @Integer (-10, 3)
+  -- NOTE: here we need to be careful not to underflow the round number
+  let roundNo' =
+        PerasRoundNo $
+          fromIntegral $
+            max 0 $
```

Comment:
Are you sure `max 0` will help for underflow? Since `0 + (-1) = maxBound` for unsigned types